blog/2021-06-29-transaction-presigning.md

```diff
@@ -0,0 +1,152 @@
+Storing the *key-offsets* within the transaction is much more space efficient because a) the values are much smaller and b) we can store them as variable length integers.
+
+The problem is that this list of key-offsets is also incorporated into the transaction's signature hash.
+
+Why is this a problem?
+Imagine two transactions, `A` and `B`:
+
+- `B` spends an output from `A` and we would like to pre-sign this spending transaction.
+- To sign `B`, we need to compute the signature hash and run the CLSAG algorithm using this hash.
+- To compute the signature hash, we need to come up with the key-offsets of our ring.
+- To compute the key-offsets for the ring, we need to know, what the output index of `A` is.
+- To know the output index of `A`, the transaction needs to be picked up by a miner and included in the blockchain.
+
+Remember that the whole point of pre-signing `B` is that we have a transaction with valid signatures **prior** to the point of signing and broadcasting `A`.
+Because the key-offsets are included in the signature hash, we can't produce it up until the output we want to spend is included in the blockchain.
+
+In summary: We cannot pre-sign transaction `B` which spends from transaction `A` before transaction `A` has been included in the blockchain.
+
+## How to achieve pre-signing on Monero?
+
+To sign a transaction using CLSAG, we don't actually need the key-offsets but the public keys these key-offsets point to!
+The public key of an output never changes, regardless of whether or not it has been included in the blockchain.
+
+To fix our pre-signing issue, we need to change how the signature hash of a transaction is being computed.
+We cannot just remove the key-offsets from the transaction.
+It is important that the signature *commits*[^2] to the ring members that were used to create it.
+At the moment though, we are *over-committing*.
+
+Instead of hashing the list of key-offsets, we can just hash the actual ring members (i.e. the public keys) instead. 
+Those remain unchanged, regardless of which outputs are already included in the blockchain.
+The key-offsets would still remain part of the transaction to retain the space-efficient look-up algorithm of the ring members.
+
+Unfortunately, this change requires a hard-fork of Monero.
+Changing how the signature hash is computed will make transactions from new clients appear invalid for an old client.
+
+## What does this mean for applications on top?
+
+It is our assumption that anything that relies on joint-outputs and pre-signing spending transactions doesn't actually work on present day Monero. 
+We haven't done an extensive analysis on what this affects but our guess is that this applies to at least anything Layer2 but likely also to other blockchain protocols developed for Monero.
+Most "interesting" blockchain protocols today work on the basis of a joint-owned output whose spending transactions have game-theoretic elements to it like learning a secret value. 
```

Comment:
:wrench: Wrong tense.
```suggestion
Most "interesting" blockchain protocols today work on the basis of a joint-owned output whose spending transactions have game-theoretic elements to them like learning a secret value. 
```

Reply:
:question: Also, is learning a secret value related to game theory? I thought that game theory in blockchain protocols had to do with leveraging the incentives of parties to perform certain actions. I would maybe just say:
```suggestion
Most "interesting" blockchain protocols today work on the basis of a joint-owned output with spending transactions which leak secrets when broadcast. 
```